File: Algorithms/078.Subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& S) {
        sort(S.begin(), S.end());
        vector<vector<int>> results;
        vector<int> current;
        subsetsHelper(results, current, S, 0);
        return results;
    }
    void subsetsHelper(vector<vector<int>>& result, vector<int>& current,
            vector<int>& S, int pos) {
        result.push_back(current);
        for (int i = pos; i < S.size(); i++) {
            current.push_back(S[i]);
            subsetsHelper(result, current, S, i+1);
            current.pop_back();
        }
    }
    /*
    vector<vector<int>> subsets(vector<int>& S) {
        sort(S.begin(), S.end());        
        int n = S.size();
        vector<vector<int>> result;
        for (int i=0; i < pow(2, n); i++) {
            vector<int> v;
            int j = i;
            int k = 0;
            while (j != 0) {
                if (j & 0x01) {
                    v.push_back(S[k]);
                }
                k ++;
                j >>= 1;
            }
            result.push_back(v);
        }
        return result;
    }
    */
};

File: Algorithms/py/078.Subsets.py
class Solution:
    # @param S, a list of integer
    # @return a list of lists of integer
    def subsets_1(self, S):
        #S.sort()
        results, current = [], []
        def _subsetsHelper(S, pos):
            results.append(current[:])
            for i in range(pos, len(S)):
                current.append(S[i])
                _subsetsHelper(S, i+1)
                current.pop()
        _subsetsHelper(S, 0)
        return results
    
    def subsets_2(self, S):
        return self.subsets_index(S, 0)
    
    def subsets_index(self, S, index):
        if index == len(S):
            return [[]]
        all_subsets = self.subsets_index(S, index+1)
        new_subsets = []
        for subset in all_subsets:
            new_subset = subset[:]
            new_subset.append(S[index])
            new_subsets.append(new_subset)
        all_subsets.extend(new_subsets)
        return all_subsets

    def subsets(self, S):
        result = []
        for i in range(0, pow(2, len(S))):
            subset = []
            j = 0
            while i != 0:
                if i & 0x01 != 0:
                    subset.append(S[j])
                i >>= 1
                j += 1
            result.append(subset)
        return result
